Add DinoV2 for semantic segmentation

kprokofi opened the pull request:
### Summary

<!--
Resolves #111 and #222.
Depends on #1000 (for series of dependent commits).

This PR introduces this capability to make the project better in this and that.

- Added this feature
- Removed that feature
- Fixed the problem #1234
-->

### How to test

<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

kprokofi committed:
added dino_v2_seg

kprokofi committed:
enabled dinov2

kprokofi committed:
fix pre-commit

kprokofi committed:
revert file back

kprokofi commented:
Need to conduct more experiments and establish hyperparameters for DINOv2 config before setting this PR in review

kprokofi committed:
fixed issues. Adapt hyperparametrs

kprokofi committed:
fixed pre-commit

kprokofi committed:
fixed mypy

kprokofi commented:
I deleted BNHead since Linear Head performed on par or even a bit better in my experiments. Freezing backbone is helpful for fine tuning task + saves training time

Final results:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


  | Pascal-VOC (cut) | kvasir-small | kvasir large | cityscapes (cut)
-- | -- | -- | -- | --
DinoV2s | 81.33 | 83.77 | 86.54 | 61.82



</body>

</html>

vinnamkim commented on src/otx/algo/segmentation/backbones/dinov2.py:
```diff
@@ -0,0 +1,54 @@
+            self.backbone.get_intermediate_layers,
+            n=out_index,
+            reshape=True,
+     )
```
```suggestion
          )
```
@harimkang 
Again, there is no code formatter in the pre-commit hook.

harimkang replied:
I'll look into it and give you some suggestions, thanks for letting me know.

vinnamkim commented on src/otx/recipe/segmentation/dino_v2_seg.yaml:
```diff
@@ -0,0 +1,130 @@
+          loss_weight: 1.0
+      pretrained: null
+      test_cfg:
+        mode: whole
```
I investigated the configuration in this notebook, https://github.com/facebookresearch/dinov2/blob/main/notebooks/semantic_segmentation.ipynb.
However, what's the reason for the difference in `test_cfg`?
![image](https://github.com/openvinotoolkit/training_extensions/assets/26541465/14ca9bde-86bc-421a-b0ed-6b316f7e6a3b)
You used `mode: whole` but the notebook used `mode: slide`.

kprokofi replied:
I aligned testing configuration with other our segmentation models. There shouldn't be any differences in final metric. I also aligned the resolution to our standard resolution that we use in Geti.

vinnamkim replied:
<img width="550" alt="image" src="https://github.com/openvinotoolkit/training_extensions/assets/26541465/15680f63-3d28-490d-b922-f4a515a9b4ce">

https://mmsegmentation.readthedocs.io/en/latest/advanced_guides/models.html#prediction-modes

I can see here that `test_cfg` really affects to the model `forward()` in the prediction. It seems being able to affect the final metric.

For VisionTransformer, input size is more important than ConvNet. This is because the input size affects to the token sequence length and it affects to the both sides, model accuracy and performance:

1) Self attention computation is `O(n^2)`, `n` is token sequence length.
2) If the pretrained model had been training for `n` tokens, but you gave `m > n` tokens after that, the position embedding for `k=n+1,...,m` might not have been learnt well.

kprokofi replied:
Okay, let me check if there will be the difference

kprokofi replied:
So, considering my experiments, I stick to current settings

kprokofi committed:
Update src/otx/algo/segmentation/backbones/dinov2.py

Co-authored-by: Vinnam Kim <vinnam.kim@intel.com>

kprokofi commented:
@vinnamkim 
I fixed seeds and set deterministic True for fair experiments. I conducted training with different settings to compare data resolution and inference methods (whole vs slide).
There is no difference in metric between the slide and the whole inference.
Also, there is almost no difference in metric for two resolutions (560 and 640). However, the model is slower with higher resolution as expected

I propose to align with our other templates and keep current settings for DINOV2

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{mso-number-format:Fixed;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Model | VOC | Cityscapes | Kvasir-small | VOC, e2e time | Cityscapes, e2e time | Kvasir-small, e2e time | AVG mIoU | AVG Time
-- | -- | -- | -- | -- | -- | -- | -- | --
slide_640 | 81.21 | 62.85 | 83.29 | 2130 | 3215 | 100 | 75.78 | 1815.00
slide_560 | 81.66 | 61.32 | 84.33 | 1712 | 2980 | 87 | 75.77 | 1593.00
whole_640 | 81.21 | 62.85 | 83.29 | 2100 | 3184 | 99 | 75.78 | 1794.33
whole_560 | 81.66 | 61.32 | 84.33 | 1695 | 2947 | 87 | 75.77 | 1576.33



</body>

</html>

kprokofi committed:
fixed pre-commit. Fixed segnext templates

kprokofi commented:
Rebased to V2 branch to resolve conflicts
